pkg/controllers/provisioning/launcher.go

```diff
@@ -38,24 +38,36 @@ import (
 )
 
 type Launcher struct {
-	Packer        *binpacking.Packer
-	KubeClient    client.Client
-	CoreV1Client  corev1.CoreV1Interface
-	CloudProvider cloudprovider.CloudProvider
+	Packer          *binpacking.Packer
+	KubeClient      client.Client
+	CoreV1Client    corev1.CoreV1Interface
+	CloudProvider   cloudprovider.CloudProvider
+	ResourceCounter *ResourceCounter
 }
 
-func (l *Launcher) Launch(ctx context.Context, schedules []*scheduling.Schedule, instanceTypes []cloudprovider.InstanceType) error {
+func (l *Launcher) Launch(ctx context.Context, schedules []*scheduling.Schedule,
+	instanceTypes []cloudprovider.InstanceType, provisioner v1alpha5.Provisioner) error {
+
 	// Pack and bind pods
 	errs := make([]error, len(schedules))
+	remainingResources, err := l.ResourceCounter.remainingResourceCounts(ctx, provisioner.Spec.Limits, provisioner.Name)
+	if err != nil {
+		return fmt.Errorf("cannot determine remaining resources, not launching instances: %w", err)
+	}
 	workqueue.ParallelizeUntil(ctx, len(schedules), len(schedules), func(index int) {
 		for _, packing := range l.Packer.Pack(ctx, schedules[index], instanceTypes) {
+			if remainingResources.isInsufficient(ctx) {
+				errs[index] = multierr.Append(errs[index], fmt.Errorf("provisioner limits exceeded"))
+				return
```

Comment:
Should we continue here? What if a pod needs CPU but not GPU and GPUs are limited.

Reply:
Ah, won't it be a part of a different schedule in that case? My understanding was that everything a part of the same packings will have identical constraints